src/main/java/vazkii/patchouli/api/data/EntryBuilder.java

```diff
@@ -0,0 +1,194 @@
+	private Boolean readByDefault;
+	private Integer sortnum;
+	private String turnin;
+	private Map<ItemStack, Integer> extraRecipeMappings;
+
+	EntryBuilder(String id, String name, String icon, CategoryBuilder parent) {
+		this.id = new ResourceLocation(parent.getId().getNamespace(), id);
+		this.name = name;
+		this.category = parent.getId().getPath();
+		this.icon = icon;
+		this.parent = parent;
+	}
+
+	EntryBuilder(String id, String name, ItemStack icon, CategoryBuilder parent) {
+		this(id, name, PatchouliAPI.instance.serializeItemStack(icon), parent);
+	}
+
+	JsonObject toJson() {
+		JsonObject json = new JsonObject();
+		json.addProperty("name", name);
+		json.addProperty("category", category);
+		json.addProperty("icon", icon);
+		JsonArray pages = new JsonArray();
+		for (AbstractPageBuilder<?> page : this.pages) {
+			pages.add(page.toJson());
+		}
+		json.add("pages", pages);
+		if (advancement != null) {
+			json.addProperty("advancement", advancement);
+		}
+		if (flag != null) {
+			json.addProperty("flag", flag);
+		}
+		if (priority != null) {
+			json.addProperty("priority", priority);
+		}
+		if (secret != null) {
+			json.addProperty("secret", secret);
+		}
+		if (readByDefault != null) {
+			json.addProperty("read_by_default", readByDefault);
+		}
+		if (sortnum != null) {
+			json.addProperty("sortnum", sortnum);
+		}
+		if (turnin != null) {
+			json.addProperty("turnin", turnin);
+		}
+		if (extraRecipeMappings != null) {
+			JsonObject mappings = new JsonObject();
+			for (Map.Entry<ItemStack, Integer> entry : extraRecipeMappings.entrySet()) {
+				mappings.addProperty(PatchouliAPI.instance.serializeItemStack(entry.getKey()), entry.getValue());
+			}
+			json.add("extra_recipe_mappings", mappings);
+		}
+		return json;
+	}
+
+	public CategoryBuilder build() {
+		return parent;
+	}
+
+	public EntryBuilder addSimpleTextPage(String text) {
+		return addPage(new TextPageBuilder(text, this)).build();
+	}
+
+	public EntryBuilder addSimpleTextPage(String text, String title) {
+		return addPage(new TextPageBuilder(text, title, this)).build();
+	}
+
+	public TextPageBuilder addTextPage(String text) {
+		return addPage(new TextPageBuilder(text, this));
+	}
+
+	public TextPageBuilder addTextPage(String text, String title) {
+		return addPage(new TextPageBuilder(text, title, this));
+	}
+
+	public ImagePageBuilder addImagePage(ResourceLocation image) {
+		return addPage(new ImagePageBuilder(image, this));
+	}
+
+	public CraftingPageBuilder addCraftingPage(ResourceLocation recipe) {
+		return addPage(new CraftingPageBuilder(recipe, this));
+	}
+
+	public SmeltingPageBuilder addSmeltingPage(ResourceLocation recipe) {
+		return addPage(new SmeltingPageBuilder(recipe, this));
+	}
+
+	public EntityPageBuilder addEntityPage(String entity) {
+		return addPage(new EntityPageBuilder(entity, this));
+	}
+
+	public EntityPageBuilder addEntityPage(ResourceLocation entity) {
+		return addPage(new EntityPageBuilder(entity.toString(), this));
```

Comment:
This might be cleaner as:
```suggestion
		return addEntityPage(entity.toString());
```